mailet/scala/src/main/scala/org/apache/james/transport/mailets/XOriginatingIpInNetwork.scala

```diff
@@ -0,0 +1,49 @@
+/** **************************************************************
+  * Licensed to the Apache Software Foundation (ASF) under one   *
+  * or more contributor license agreements.  See the NOTICE file *
+  * distributed with this work for additional information        *
+  * regarding copyright ownership.  The ASF licenses this file   *
+  * to you under the Apache License, Version 2.0 (the            *
+  * "License"); you may not use this file except in compliance   *
+  * with the License.  You may obtain a copy of the License at   *
+  * *
+  * http://www.apache.org/licenses/LICENSE-2.0                 *
+  * *
+  * Unless required by applicable law or agreed to in writing,   *
+  * software distributed under the License is distributed on an  *
+  * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+  * KIND, either express or implied.  See the License for the    *
+  * specific language governing permissions and limitations      *
+  * under the License.                                           *
+  * ***************************************************************/
+
+package org.apache.james.transport.mailets
+
+import java.util.Collection
+
+import org.apache.james.core.MailAddress
+import org.apache.james.transport.matchers.AbstractNetworkMatcher
+import org.apache.mailet.Mail
+
+object XOriginatingIpInNetwork {
+  val X_ORIGINATING_IP: String = "X-Originating-IP"
+}
+
+class XOriginatingIpInNetwork extends AbstractNetworkMatcher {
+  override def `match`(mail: Mail): Collection[MailAddress] = {
+    matchOnOriginatingAddr(mail).orNull
+  }
+
+  def matchOnOriginatingAddr(mail: Mail): Option[Collection[MailAddress]] = {
+    Option(mail.getMessage
+      .getHeader(XOriginatingIpInNetwork.X_ORIGINATING_IP))
+      .flatMap(_.headOption)
```

Comment:
Maybe I should only check the last one, relying on declarative headers which may have been added by external clients doesn't feel very safe as it is very easy to record fake information ins such headers. In our use case because of how the infrastructure is designed there can only be one such header and we check both the direct sender and the X_ORIGINATING_IP. 
I chose not to implement a complex behaviour which would require more extensive configuration.